Pad arrays

drandykass opened the pull request:
This is an incomplete pull request for the addition of array padding and unpadding routines to gridder.

Some talking points:
--The padding function expects the array to be of the same dimension as intended--that is, no 'flattened' arrays.
--The unpadding function currently is set up to return unpadded coordinate vectors.  However, I think that may be completely unnecessary.
--Still working on cleaning up the docs, the actual code, and need to create the test functions.

Update 03-12-2015 (@leouieda): Added the checklist below.

## Checklist:

- [x] Make tests for new code
- [x] Create/update docstrings
- [x] Include relevant equations and citations in docstrings
- [x] Code follows PEP8 style conventions
- [x] Code and docs have been spellchecked
- [x] Include new dependencies in docs, requirements.txt, README, and .travis.yml
- [x] Documentation builds properly
- [x] All tests pass
- [x] Can be merged
- [x] Changelog entry (leave for last)
- [x] Firt-time contributor? Add yourself to `doc/contributors.rst` (leave for last)

leouieda commented:
Thanks @drandykass! 

A few initial comments:

> The padding function expects the array to be of the same dimension as intended--that is, no 'flattened' arrays.

That's reasonable. The padding only makes sense for gridded data anyway. This should be explicit in the function docstring.

> The unpadding function currently is set up to return unpadded coordinate vectors. However, I think that may be completely unnecessary.

I agree that it's redundant since the original arrays should be available from the start. It would make the function simpler if it takes only the padded array. Plus, you could unpad the coordinates by calling this function, right?

> Still working on cleaning up the docs, the actual code, and need to create the test functions.

OK, let me know if you need help with that.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,244 @@ def cut(x, y, scalars, area):
               and y[i] >= ymin and y[i] <= ymax]
     return [x[inside], y[inside], [s[inside] for s in scalars]]
+
+def pad_array(a, xy=None, np=None, padtype='OddReflectionTaper'):
```
@drandykass it's better to avoid `np` for variable names since that's usually what people import numpy as. We've been changing this import in Fatiando lately to make code a bit nicer to read. So if you want to change the `import numpy`  to `import numpy as np` in this module, please feel free. You'd need to replace all calls to `numpy` with `np` though.

drandykass replied:
Good point.  I'll make the appropriate modifications (I'll leave numpy as
is and just change the variable name).

On Thu, Dec 3, 2015 at 10:26 AM, Leonardo Uieda <notifications@github.com>
wrote:

> In fatiando/gridder.py
> <https://github.com/fatiando/fatiando/pull/239#discussion_r46582582>:
>
> > @@ -439,3 +444,244 @@ def cut(x, y, scalars, area):
> >                if x[i] >= xmin and x[i] <= xmax
> >                and y[i] >= ymin and y[i] <= ymax]
> >      return [x[inside], y[inside], [s[inside] for s in scalars]]
> > +
> > +def pad_array(a, xy=None, np=None, padtype='OddReflectionTaper'):
>
> @drandykass <https://github.com/drandykass> it's better to avoid np for
> variable names since that's usually what people import numpy as. We've been
> changing this import in Fatiando lately to make code a bit nicer to read.
> So if you want to change the import numpy to import numpy as np in this
> module, please feel free. You'd need to replace all calls to numpy with np
> though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239/files#r46582582>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

drandykass commented:
OK, I think she's ready to go--tests fine on my system.  Pending review by others, the only thing I think that's left on the checklist is a changelog entry and contributor addition.  Let me know if you find anything!

leouieda commented:
@drandykass great! That was fast.

I took a quick look at the code and I have a few suggestions. Look into this with more care later on this week and get back to you.

drandykass commented:
Great.  Let me know if you expect the function call structure to change.
I'm going to finish up the fourier domain calculation functions, which will
use these.

On Thu, Dec 3, 2015 at 1:49 PM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> great! That was fast.
>
> I took a quick look at the code and I have a few suggestions. Look into
> this with more care later on this week and get back to you.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-161780475>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented on cookbook/grid_pad.py:
```diff
@@ -0,0 +1,79 @@
+pads = []
+xy = np.zeros((len(x),2))
+xy[:,0] = x[:]
+xy[:,1] = y[:]
```
@drandykass the handling of the coordinate arrays seems a bit too complicated. Wouldn't it be better to pass along (and return) a list of coordinate arrays? That would also avoid the memory copying involved in creating a new 2D array (the list holds only references to the arrays, not their values).

drandykass replied:
Sure, a list is fine.  There was a reason I did it this way--couldn't tell
you what that reason was now though...

It returns a list of coordinate arrays so no reason to not expect a list
coming in.  Wonder why I did it that way?

Anyway, I'll start a \todo list and add this to it.

On Fri, Dec 4, 2015 at 12:04 PM, Leonardo Uieda <notifications@github.com>
wrote:

> In cookbook/grid_pad.py
> <https://github.com/fatiando/fatiando/pull/239#discussion_r46718529>:
>
> > +import numpy as np
> > +
> > +# Generate gridded data
> > +shape = (101,172)
> > +x,y,z = gridder.regular((-5000, 5000, -5000, 5000), shape, z=-150)
> > +model = [mesher.Prism(-4000, -3000, -4000, -3000, 0, 2000, {'density':1000}),
> > +        mesher.Prism(-1000, 1000, -1000, 1000, 0, 2000, {'density':-900}),
> > +        mesher.Prism(2000, 4000, 3000, 4000, 0, 2000, {'density':1300})]
> > +gz = prism.gz(x, y, z, model)
> > +gz = gz.reshape(shape)
> > +
> > +# Pad arrays with all the padding options
> > +pads = []
> > +xy = np.zeros((len(x),2))
> > +xy[:,0] = x[:]
> > +xy[:,1] = y[:]
>
> @drandykass <https://github.com/drandykass> the handling of the
> coordinate arrays seems a bit too complicated. Wouldn't it be better to
> pass along (and return) a list of coordinate arrays? That would also avoid
> the memory copying involved in creating a new 2D array (the list holds only
> references to the arrays, not their values).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239/files#r46718529>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda replied:
:+1: that happens a lot to me. One of the advantages of writing long git commit messages (though it's easier said than done).

leouieda commented on fatiando/gridder.py:
```diff
@@ -80,7 +85,7 @@ def load_surfer(fname, fmt='ascii'):
             ny, nx = [int(s) for s in ftext.readline().split()]
             shape = (nx, ny)
-            # Read the min/max value of columns/longitue (y direction)
+            # Read the min/max value of columns/longitude (y direction)
```
Thanks for the typo fix!

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+
+        [ OddReflectionTaper | OddReflection | Reflection | value | LinTaper
+        | edge | mean ]
+            OddReflectionTaper - Generates odd reflection then tapers to the 
```
I think it would be more consistent to have all names lowercase. That would also remove the need to the `.lower()` calls below.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+
+            LinTaper - Linearly tapers to the mean
+
+            value - numeric value
```
Could you clarify the explanation here? It's not that clear that one should pass `'0'` and not `'value'` for this option. I was a bit confused until I saw the cookbook. Maybe add the comments in the recipe that you use to explain this.

drandykass replied:
Lower case names: sure.  I'll leave in the .lower() though so that it
catches if someone has caps lock or whatever on.

Clarify explanation: Can do.

On Fri, Dec 4, 2015 at 12:13 PM, Leonardo Uieda <notifications@github.com>
wrote:

> In fatiando/gridder.py
> <https://github.com/fatiando/fatiando/pull/239#discussion_r46719655>:
>
> > +        dimension
> > +    * padtype : string (optional)
> > +        String describing what to pad the new values with. Options:
> > +
> > +        [ OddReflectionTaper | OddReflection | Reflection | value | LinTaper
> > +        | edge | mean ]
> > +            OddReflectionTaper - Generates odd reflection then tapers to the
> > +            mean using a cosine function (Default)
> > +
> > +            OddReflection - Pads with the odd reflection, with no taper
> > +
> > +            Reflection - Pads with simple reflection
> > +
> > +            LinTaper - Linearly tapers to the mean
> > +
> > +            value - numeric value
>
> Could you clarify the explanation here? It's not that clear that one
> should pass '0' and not 'value' for this option. I was a bit confused
> until I saw the cookbook. Maybe add the comments in the recipe that you use
> to explain this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239/files#r46719655>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda replied:
>  I'll leave in the .lower() though so that it catches if someone has caps lock or whatever on.

I'd think that would be a good reason to remove them. Kind of like how good password fields warn you that you caps on. But I'm okay with leaving them.

PS: You seen to be replying from the e-mails. I find it easier to view the comments on the website as the code appears right above them.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+        dimension.  This is effectively a concatenated xp,yp, etc...
+    * npd : tuple (optional)
+        Optional tuple containing the total number of desired elements in each
+        dimension
```
Could you add some text saying what happens if `npd` is `None`? I see you put it in a comment but the user wouldn't be able to see that.

leouieda replied:
> Optional tuple containing the total number of desired elements in each dimension.

Another way to say this with a more "numpy" feel would be: "Desired shape of the new padded array". Maybe put this in as well as the current description? This would benefit both users familiar with numpy and those a bit new still.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+        Returns None if xy not provided
+    * nps : list
+        List of tuples containing the number of elements padded onto each 
+        dimension.
```
Do you really need to return this? This argument is a bit confusing. I see that this is used to unpad the array later but this information is contained in the shape of the padded array, right?

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+
+    # Test to make sure padtype is valid
+    padopts = ['','oddreflectiontaper','oddreflection', 'reflection',
+            'lintaper', 'edge', 'value', 'mean']
```
What happens if `padtype=''` (empty string)? I'm curious because you added it as a valid option.

drandykass replied:
Returning nps:
It's a luxury--it would be encoded in the padded array, although if you had
an odd number of padding elements, you would need to know which side had
the extra element.  You could find that out by comparing your coordinate
vectors, or by knowing which side the extra element is padded on.  I
personally think it makes life easier, if a tad more messy.

padtype = ' '  Oops, that should not be a valid option.  The default is
provided in the function definition, so there really should not be an
option here.  I'll fix.

Other previous comments: Agreed.

On Fri, Dec 4, 2015 at 12:20 PM, Leonardo Uieda <notifications@github.com>
wrote:

> In fatiando/gridder.py
> <https://github.com/fatiando/fatiando/pull/239#discussion_r46720655>:
>
> > +    Returns:
> > +
> > +    * ap : numpy array
> > +        Padded array. The array core is a deep copy of the original array
> > +    * cp : list
> > +        List of coordinate arrays containing the extrapolated coordinate values
> > +        Returns None if xy not provided
> > +    * nps : list
> > +        List of tuples containing the number of elements padded onto each
> > +        dimension.
> > +
> > +    """
> > +
> > +    # Test to make sure padtype is valid
> > +    padopts = ['','oddreflectiontaper','oddreflection', 'reflection',
> > +            'lintaper', 'edge', 'value', 'mean']
>
> What happens if padtype='' (empty string)? I'm curious because you added
> it as a valid option.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239/files#r46720655>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda replied:
> It's a luxury--it would be encoded in the padded array, although if you had an odd number of padding elements, you would need to know which side had the extra element. You could find that out by comparing your coordinate vectors, or by knowing which side the extra element is padded on. I personally think it makes life easier, if a tad more messy.

You're completely right! I hadn't thought of odd number padding. That's perfect the way it is then. 

Maybe you could add a small example of how that argument looks? That might help clarify. It could even be a doctest like [this one](https://github.com/fatiando/fatiando/blob/master/fatiando/gridder.py#L141). That would help documenting and testing at the same time.

drandykass replied:
Could you clarify what you mean by adding example of how the argument looks?  I'm not sure what I'm looking at in the example.  It just returns a list with each element containing a tuple with the number of cells padded on the lower and upper sides, respectively.  Dimension order is the same as what was input.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+                raise ValueError(et)
+            else:
+                npt = npd
+        for ii in numpy.arange(len(npt)):
```
Be careful of `numpy.arange` for looping over an index because it  might not produce integers (that would break the indexing code below). It's always better to use `range` or `xrange`.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+    #  [ ] if statements below - Finally add an elif to the if statements
+    #                            below. Add a descriptive comment.
+    if padtype.lower() not in padopts and not _is_number(padtype):
+        raise ValueError('gridder.pad_array: Pad Type not understood')
```
There is no need to add `gridder.pad_array: ` to the exception. That information is already included in the traceback.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+        for ii in numpy.arange(len(npt)):
+            if npt[ii] <= a.shape[ii]:
+                # How are you supposed to properly assign a multi-line string?
+                et1 = 'gridder.pad_array: Desired padding is less than array '
```
First, you can remove `gridder.pad_array: ` from the message. 

Also, this might look better:
    
    raise ValueError(
        "Very long error message " + 
        "that I'll split into multiple lines " +
        "using + operations.")

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+    if _is_number(padtype):
+        # Pad with value
+        ap = numpy.pad(numpy.copy(a), nps, mode='constant',
+            constant_values = float(padtype))
```
Doesn't `numpy.pad` already produce a copy? I would think so. Usually the inplace operations are done via array methods.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+                rp=nps[ii][1])
+        ap += m
+    if xy is not None:
+        cp = _padcoords(xy, a.shape, nps)
```
Hum. This makes me think that maybe it would be simpler to just make `_padcoords` into a public `pad_coords` function that does the extending of the coordinate arrays. That would greatly simplify the code, calling and returning of this function.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+    Effectively, this is a complement to gridder.cut for when you already 
+    know the number of elements to remove.
+
+    .. note: Unlike pad_array, this returns a slice of the input array.
```
If you replace `Unlike pad_array,` with 

    Unlike :func:`~fatiando.gridder.pad_array`,

Sphinx will replace that with a link the `pad_array` function docs.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +444,269 @@ def cut(x, y, scalars, area):
+    coords = []
+    d = []
+    coordspad = []
+    for ii in numpy.arange(len(s)):
```
Is `ii` something from Matlab or Fortran to avoid conflict with complex numbers? If so, that wouldn't be a problem in Python.

leouieda commented on test/test_gridder_padarray.py:
```diff
@@ -0,0 +1,83 @@
+        mesher.Prism(-1000,1000,-1000,1000,0,2000,{'density':-900}),
+        mesher.Prism(2000,4000,3000,4000,0,2000,{'density':1300})]
+    gz = prism.gz(x,y,z,model)
+    gz = gz.reshape(s)
```
The tests are great! Very comprehensive!

One this I'd change is this call to the forward modeling code. I don't think it's very good to rely on such complex code for a unit test. It would be much better to use a simpler function, like `x**2 + y**3` or random numbers like below.

drandykass replied:
Do you have objection to using the Rosenbrock function as an example?

leouieda commented on test/test_gridder_padarray.py:
```diff
@@ -0,0 +1,83 @@
+
+def test_pad_and_unpad_equal_1d():
+    'gridder.pad_array and subsequent .unpad_array gives original array: 1D'
+    x = np.random.rand(21)
```
It's better to use a defined random seed in unit tests to make sure that you can always repeat a failing test later on.

leouieda commented on test/test_gridder_padarray.py:
```diff
@@ -0,0 +1,83 @@
+    gz = gz.reshape(s)
+    gpad,_,nps = gridder.pad_array(gz)
+    gunpad = gridder.unpad_array(gpad,nps)
+    assert_almost(gunpad,gz)
```
Could you add a `for` loop here that checks all available padding options?

leouieda commented on test/test_gridder_padarray.py:
```diff
@@ -0,0 +1,83 @@
+        mesher.Prism(-1000,1000,-1000,1000,0,2000,{'density':-900}),
+        mesher.Prism(2000,4000,3000,4000,0,2000,{'density':1300})]
+    gz = prism.gz(x,y,z,model)
+    gz = gz.reshape(s)
```
Same thing here about the use of  prisms.

leouieda commented:
One last thing, for some reason the PEP8 unit test is not catching the format errors as it should. You can run `make pep8` to get a full report. It will tell what is off and what you have to do to fix it.

drandykass commented:
Yes, I was replying to the emails--gmail was showing me the code though, so it was working ok.

I will apply all these suggestions, with a bit of a discussion on a few.
--The ii is an old habit to simply distinguish from imaginary numbers.  Actually, any repeated variable in my code is guaranteed to be an indexing variable (just a personal style thing)

--I hear you on the pad_coords.  I can certainly make that public.  Do you think there's still value in having it be called from pad_array though, for simplicity to the end user?  I'm ok either way.
Option a: the user sends their array and coordinate arrays to pad_array, though pad_coords is public for whomever wants to use it
Option b: the user sends an array to pad_array, and their coordinate vectors to pad_coords

--I'm not sure if numpy.pad produces a copy.  I'll check and do what is appropriate (let's not make 2 copies!)

--Other than that, I agree 100% on everything else.  I'll apply those changes, write up a changelog, and get that submitted this weekend sometime, and we can iterate and see if there's anything else.  Thanks for that detailed review!

leouieda commented:
> The ii is an old habit to simply distinguish from imaginary numbers. Actually, any repeated variable in my code is guaranteed to be an indexing variable (just a personal style thing)

OK, not a problem. Just found it curious and I think I've seen it before (I think Valeria uses that sometimes). 

> I hear you on the pad_coords. I can certainly make that public. Do you think there's still value in having it be called from pad_array though, for simplicity to the end user? I'm ok either way. Option a: the user sends their array and coordinate arrays to pad_array, though pad_coords is public for whomever wants to use it. Option b: the user sends an array to pad_array, and their coordinate vectors to pad_coords.

I tend towards option b: `pad_array` only acts on the data array and the user calls `pad_coords` if he/she wants. That way `pad_array` doesn't need to know about coordinates at all and the code doesn't have to deal with optional arguments.

> I'm not sure if numpy.pad produces a copy. I'll check and do what is appropriate (let's not make 2 copies!)

I ran some tests here and it does already make a copy. 

> Other than that, I agree 100% on everything else. I'll apply those changes, write up a changelog, and get that submitted this weekend sometime, and we can iterate and see if there's anything else. Thanks for that detailed review!

No problem! Reviewing is something that I really want to establish as a habit for Fatiando. You can pay me back with some reviews in the future :wink: Post a comment here when you push your updates so that I get a notification.

leouieda commented:
@drandykass I just merged in some changes to fix a few problems with the documentation build.  Could you please merge these into your branch?

drandykass commented:
Yes. I haven't addressed your comments yet either...trying to write an exam
before this afternoon!

Sent from mobile device. Please excuse errors.
On Dec 9, 2015 7:32 AM, "Leonardo Uieda" <notifications@github.com> wrote:

> @drandykass <https://github.com/drandykass> I just merged in some changes
> to fix a few problems with the documentation build. Could you please merge
> these into your branch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-163263010>.
>

drandykass commented:
I just added 2 inline comments above; I wasn't sure if it would alert you.  I've applied most of your changes except for the few questions I had above, and I'll finish up the rest soon.

drandykass commented:
Other than the clarification I need above, everything is done.  All your comments have been addressed.  Pep8 compliance is met, docstrings seem to be in good shape. Added myself to contributors, and made an entry in the changelog.  Let me know if there's anything else.

I did go ahead and use the rosenbrock function in the unit tests.  I only did that since it's such a commonly used example function in optimization theory testing.  If you'd prefer something even simpler, I totally understand--it's a 20 second change.

I am not passing the travis build, but it's passing on my machine.  It's the checking of the 2D padded array that's failing on here, but seems fine on my computer.  Any help would be appreciated.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +445,281 @@ def cut(x, y, scalars, area):
+    return b
+
+
+def padcoords(xy, s, nps):
```
Could you rename this `pad_coords`, just to be in the same theme as `pad_arrays` and `unpad_arrays`? (That's my OCD talking...)

drandykass replied:
Ah geez, you slavedriver!  Sure thing, no problem.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +445,281 @@ def cut(x, y, scalars, area):
+
+    * xy : List
+        List of arrays of coordinates
+    * s : tuple
```
Also, could you rename this argument to `shape`? That's the standard numpy convention and is used everywhere, so it would be nice to maintain consistency.

drandykass replied:
I can do that.  The reason I had changed it to 's' was to not overload the variable name with the a.shape property.  I had noticed you used shape everywhere else.

That's fine though.  Python won't mind so neither will I.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +445,281 @@ def cut(x, y, scalars, area):
+        if type(xy) is numpy.ndarray:
+            coords.append(xy)
+        elif type(xy) is list and len(s) > 1:
+            coords.append(xy[ii].reshape(s).transpose().take(0, axis=ii))
```
type checking in Python is not a very good idea because it invalidates all the possible [duck typing](https://en.wikipedia.org/wiki/Duck_typing) that makes Python so flexible. For example, I could not  pass an [xray](http://xray.readthedocs.org/en/stable/) array to this function but there is no reason why it shouldn't work.

A better option in this case is to ask for a list of 1d arrays and assume that that is what you get.  So you can do:

    for array in xy:

drandykass replied:
Would it still be ok to do a type check to make sure a list is passed in then?  Then if what is passed in is not a list, it just dumps whatever that is (xray or whatever) into a single element list.

It's not strictly necessary by any means.  I'm just so into checking arguments.  If someone isn't thinking and passes a single vector in (which makes the most sense if working in 1D), I'd like it to work.  If that's outside the scope of this though, that's fine.

leouieda commented:
Rosenbrock is perfect, no problem! Seems almost ready. I just had a few small comments. One thing that is missing from the docstrings that would be great to have a examples (as doctests). See the [gridder.regular](http://www.fatiando.org/api/gridder.html#fatiando.gridder.regular) docstring for an example of how this is done. The examples should be simple applications of the function on small input (something that you could `print` in a few lines). Maybe something like this for `pad_arrays`:

    >>> import numpy as np
    >>> data = np.array([1, 2, 3, 3, 2, 1])
    >>> pad_array(data)  # Pads to the nearest power of 2
    array([ NUMBERS GO HERE ])
    
If you could add a simple example like this for the most common uses (1D and 2D arrays, different options) it  would make the functions much easier to understand. I always go to the examples in the numpy docs and usually only understand the function because of them.

leouieda commented:
Oh, seems I missed a few inline comments before. It seems those go away when a commit changes the line that they were inserted.

drandykass commented:
Oh, I see what you mean by the examples as doctests now.  Sure, I can add those.

drandykass commented:
OK, I addressed your changes.

padcoords is now pad_coords
s -> shape
Added docstring examples
Removed the type checking for a numpy array.  As it is now, though, it still checks to see if it's NOT a list.  If it's not a list, the program assumes that the user inadvertently sent a vector (instead of a single element list containing a vector), creates a list accordingly, and then moves on.  I was going to add an 'isinstance' check to make sure whatever they put in was iterable, but figured that was overkill.

leouieda commented on fatiando/gridder.py:
```diff
@@ -439,3 +445,325 @@ def cut(x, y, scalars, area):
+        >>> xp
+        array([-20., -15., -10.,  -5.,   0.,   5.,  10.,  15.])
+        >>> yp
+        array([-24., -20., -16., -12.,  -8.,  -4.,   0.,   4.])
```
Does `pad_coords` always return the arrays with the `meshgrid` applied? If so, then we're leaving it up to the user to apply the `meshgrid`? This sounds like something that will generate a lot of emails on the mailing list :wink: Getting the order of arguments for meshgrid is very confusing so maybe it would be best to return the flattened arrays after meshgrid. That way the user can treat the new arrays just like the old ones.

leouieda commented:
@drandykass this is almost over, I promise!

drandykass commented:
No, pad_coords returns the 1D arrays.  It's up to the user to create a
meshgrid.

In order for the output (ynew) to equal the input,

[Y,X] = meshgrid(ynew,xnew)
y = Y.ravel()
x = X.ravel()

The only issue with creating the meshgrids is that I no longer know how to
make it general for any dimension.  Thoughts?

On Thu, Dec 17, 2015 at 10:56 AM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> this is almost over, I
> promise!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-165530705>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
Does it need to be general for any dimensions? I understand `pad_array` being for n-dimesional arrays but maybe `pad_coords` could be for just 2D? Not sure.

It's important not to get too caught up on trying to predict every future use case. We'll never be able to and might over complicate things because of some feature that will never be used.

leouieda commented:
The design is much harder than the coding.

leouieda commented:
@drandykass maybe for now we just keep the code as is but add a note and example in the docstring about how to get the arrays with meshgrid. Then we keep the flexibility but keep the reminder of how to do this properly.

drandykass commented:
This is your show; we'll do what you prefer.  I think the pad_coords needs
to work for at least 1 and 2 dimension.  Anything extra is cake.

On Thu, Dec 17, 2015 at 12:03 PM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> maybe for now we just keep
> the code as is but add a note and example in the docstring about how to get
> the arrays with meshgrid. Then we keep the flexibility but keep the
> reminder of how to do this properly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-165549715>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
>This is your show; we'll do what you prefer. 

It's [our show](https://github.com/fatiando/fatiando/graphs/contributors) now! I really appreciate the inputs and this is your code so you know it best. I just think of how this fits in with the rest.

> I think the pad_coords needs to work for at least 1 and 2 dimension. Anything extra is cake.

OK, I forgot 1D. So it's probably better to leave it as is but provide instructions for doing the meshgrid in 2D.

drandykass commented:
OK, I'll add instructions and ponder on a way to do arbitrary dimension
with a branch in the future.  Might just need some logic to do 1D, 2D, or
other.

On Thu, Dec 17, 2015 at 12:51 PM, Leonardo Uieda <notifications@github.com>
wrote:

> This is your show; we'll do what you prefer.
>
> It's our show <https://github.com/fatiando/fatiando/graphs/contributors>
> now! I really appreciate the inputs and this is your code so you know it
> best. I just think of how this fits in with the rest.
>
> I think the pad_coords needs to work for at least 1 and 2 dimension.
> Anything extra is cake.
>
> OK, I forgot 1D. So it's probably better to leave it as is but provide
> instructions for doing the meshgrid in 2D.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-165563681>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
I'm all for leaving that for the future

drandykass commented:
OK, I added in an example and some text on how to convert the output from pad_coords to the flattened meshgrid-style that is used elsewhere.  That was surprisingly hard to articulate.  Passes pep8, docstring, and unittests.

Anything else?  I don't mind this iterative process--I'm just worried about annoying you with what is effectively a very simple routine.

drandykass commented:
Oh, hey, got how to have it return the flattened array for an arbitrary dimension.  

    xylist = [x, y]    %this is how it comes in
    xytuple = tuple(xylist)
    M = np.meshgrid(*[a for a in xytuple[::-1]])
    XY = []
    for a in M:
        XY.append(a.ravel())
    return XY

I can either implement this now or wait for another update.  Up to you.  Just wanted to write it down here so I didn't forget.

leouieda commented:
@drandykass happy new year!

I've had a look and everything seems perfect now. I noticed a warning in `pad_arrays` that was probably left behind from the split to `pad_coords` (https://github.com/fatiando/fatiando/pull/239/files#diff-eb344a12bd50dc1fcfecd68f5c24e1ebR468). You're having merge conflicts because of the contributors and changelog files from #190 I merged earlier today. 

Hope I can do things in a more timely manner this year.

drandykass commented:
No problem!  Since they're super easy, should I just nuke my copy of the
files, then update everything, make the changes again, and commit?

On Tue, Feb 2, 2016 at 10:22 AM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> happy new year!
>
> I've had a look and everything seems perfect now. I noticed a warning in
> pad_arrays that was probably left behind from the split to pad_coords (
> https://github.com/fatiando/fatiando/pull/239/files#diff-eb344a12bd50dc1fcfecd68f5c24e1ebR468).
> You're having merge conflicts because of the contributors and changelog
> files from #190 <https://github.com/fatiando/fatiando/pull/190> I merged
> earlier today.
>
> Hope I can do things in a more timely manner this year.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-178698749>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
Another option is to merge in the `master` branch from  fatiando/fatiando and resolve the conflicts.

drandykass commented:
Hm, which option is least likely to cause me to break everything...  :p

On Tue, Feb 2, 2016 at 10:36 AM, Leonardo Uieda <notifications@github.com>
wrote:

> Another option is to merge in the master branch from fatiando/fatiando
> and resolve the conflicts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-178705972>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
Maybe try merging master into your branch first. You'll have to pull in the
changes from here first. Resolving the conflicts is not very difficult for
simple things like this.

It's very unlikely that you could mess everything up either way. You'd have
force push in order to erase what is on github. So you can try what you
like on your clone. Go wild! :)
On Feb 2, 2016 15:38, "Andy Kass" <notifications@github.com> wrote:

> Hm, which option is least likely to cause me to break everything... :p
>
> On Tue, Feb 2, 2016 at 10:36 AM, Leonardo Uieda <notifications@github.com>
> wrote:
>
> > Another option is to merge in the master branch from fatiando/fatiando
> > and resolve the conflicts.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/fatiando/fatiando/pull/239#issuecomment-178705972>.
> >
>
>
>
> --
>
> ------------------------------------------------------------------------------------
> Dum inter homines sumus, colamus humanitatem.
>
> ------------------------------------------------------------------------------------
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-178706768>.
>

leouieda commented:
@drandykass have you been able to make any progress on this? I've been doing some cleaning of stale PRs and just wanted to check.

drandykass commented:
I was just going to email you.  I think it's something super simple like
the change log was out of order.  Honestly, I'm so far removed at this
point I'm not sure how to fix it!  Ajude-me!

On Thu, Apr 28, 2016 at 12:41 PM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> have you been able to make
> any progress on this? I've been doing some cleaning of stale PRs and just
> wanted to check.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-215523368>
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
OK, this might have gotten a bit more complicated after the 0.4 release. Will probably require some more git-fu. Can you give me write access to your fork of fatiando so I can try to fix this? You can do this by going to your fork on Github -> settings -> collaborators

drandykass commented:
I believe it's done.  Sorry my git-fu isn't up to par!

On Thu, Apr 28, 2016 at 1:19 PM, Leonardo Uieda <notifications@github.com>
wrote:

> OK, this might have gotten a bit more complicated after the 0.4 release.
> Will probably require some more git-fu. Can you give me write access to
> your fork of fatiando so I can try to fix this? You can do this by going to
> your fork on Github -> settings -> collaborators
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#issuecomment-215534020>
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
I get the notification from github. I'll get on this some time this week.

Well, git fu is only through extensive practice of goggle fu. I'd be frightened to see how many times I've googled "git undo last commit".

leouieda commented:
@drandykass I think I've got things working. For the record, what I did was:

1. Got a clone of your fork
2. Registered fatiando/fatiando as a remote option:

        git remote add upstread git@github.com:fatiando/fatiando.git

3. Pull in changes from upstream (fatiando/fatiando) to your master branch:

        git checkout master
        git pull upstream master

4. Update the master branch on your fork:

        git push origin master

5. Rebase your branch on the current master branch. This applies the commits
   from your branch on top of the current master. What happens is that this ends
   up rewriting all your commits (keeping same message etc, just the hash
   changes).

        git checkout pad_arrays
        git rebase master

6. Fix all the merge conflicts that appeared (only the changelog and
   contributors).
7. Force push to you branch the new commits, overwriting the old ones:

        git push -f origin pad_arrays

leouieda commented:
Now everything is up dated and passing tests. So I'll happily merge this! Congrats and thank you!

drandykass commented:
Thanks for all your help and hand-holding!  Now I need the fourier forward
modeling done!

On Thu, May 5, 2016 at 6:12 PM, Leonardo Uieda <notifications@github.com>
wrote:

> Merged #239 <https://github.com/fatiando/fatiando/pull/239>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/239#event-652122067>
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

leouieda commented:
Not a problem! One thing I figured out these past 2 years as a prof is that I love teaching.

Welcome (officially) to the contributors list! http://www.fatiando.org/dev/contributors.html
